samza-core/src/main/java/org/apache/samza/execution/JobNodeConfigureGenerator.java

```diff
@@ -0,0 +1,340 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.execution;
+
+import com.google.common.base.Joiner;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.samza.config.Config;
+import org.apache.samza.config.JobConfig;
+import org.apache.samza.config.MapConfig;
+import org.apache.samza.config.SerializerConfig;
+import org.apache.samza.config.StorageConfig;
+import org.apache.samza.config.StreamConfig;
+import org.apache.samza.config.TaskConfig;
+import org.apache.samza.config.TaskConfigJava;
+import org.apache.samza.operators.spec.JoinOperatorSpec;
+import org.apache.samza.operators.spec.OperatorSpec;
+import org.apache.samza.operators.spec.StatefulOperatorSpec;
+import org.apache.samza.operators.spec.StoreDescriptor;
+import org.apache.samza.operators.spec.WindowOperatorSpec;
+import org.apache.samza.serializers.KVSerde;
+import org.apache.samza.serializers.Serde;
+import org.apache.samza.serializers.SerializableSerde;
+import org.apache.samza.table.TableConfigGenerator;
+import org.apache.samza.table.TableSpec;
+import org.apache.samza.util.MathUtil;
+import org.apache.samza.util.StreamUtil;
+import org.apache.samza.util.Util;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This class provides methods to generate configuration for a {@link JobNode}
+ */
+/* package private */ class JobNodeConfigureGenerator {
+
+  private static final Logger LOG = LoggerFactory.getLogger(JobNodeConfigureGenerator.class);
+
+  static final String CONFIG_JOB_PREFIX = "jobs.%s.";
+  static final String CONFIG_INTERNAL_EXECUTION_PLAN = "samza.internal.execution.plan";
+
+  JobNodeConfigureGenerator() {
+  }
+
+  static Config mergeJobConfig(Config originalConfig, Config generatedConfig) {
+    JobConfig jobConfig = new JobConfig(originalConfig);
+    String jobId = JobNode.createId(jobConfig.getName().get(), jobConfig.getJobId());
+    return new JobConfig(Util.rewriteConfig(extractScopedConfig(originalConfig, generatedConfig,
+        String.format(CONFIG_JOB_PREFIX, jobId))));
+  }
+
+  JobConfig generateJobConfig(JobNode jobNode, String executionPlanJson) {
```

Comment:
Do you know why the executionPlanJson needs to be passed all the way through to here? Could this be added to the config by the runner/planner itself?

Reply:
So, this json string can not be generated until we create the JobGraph in the planner. And the reason to pass it to each JobNodeConfigurationGenerator is to allow the DAG to be displaced by each job, not just once per application.